s3dol/base.py

```diff
@@ -0,0 +1,227 @@
+"""S3Dol with a dict-like interface for AWS S3 buckets.
+Supports relative paths to access files under a prefix ending with slash.
+Hints added to help setup credentials.
+
+bucket = S3Dol()[profile][bucket_name]
+bucket['level1/level2/test-key'] == bucket['level1/']['level2/']['test-key']
+"""
+from dataclasses import dataclass
+import functools
+import os
+from typing import Union
+import warnings
+
+import boto3
+import dol
+
+
+class S3DolException(Exception):
+    pass
+
+
+noCredentialsFound = S3DolException(
+    'No AWS credentials found. Configure your AWS credentials as environment variables or in ~/.aws/credentials. '
+    'See https://github.com/i2mint/s3dol/#set-up-credentials for more information.'
+)
+
+
+def get_aws_credentials():
+    aws_access_key = os.environ.get('AWS_ACCESS_KEY_ID')
+    aws_secret_key = os.environ.get('AWS_SECRET_ACCESS_KEY')
+    aws_region = os.environ.get('AWS_DEFAULT_REGION')
+    aws_session_token = os.environ.get('AWS_SESSION_TOKEN')
+
+    return {
+        'aws_access_key_id': aws_access_key,
+        'aws_secret_access_key': aws_secret_key,
+        'region_name': aws_region,
+        'aws_session_token': aws_session_token,
+    }
```

Comment:
Config should not be part of s3dol. It is a separate concern. We want to pass the credentials as parameter to the stores (as we do in azuredol)

Reply:
These configs exist to stay aligned with awscli and boto3 which use these by default. And also satisfy the "No-requried arguments" of https://github.com/i2mint/s3dol/issues/2

For our use case, the credentials can also be passed like this but also with any other aws parameters enabled by boto3.  Example in [my branch of platform_poc](https://github.com/otosense/platform_poc/blob/9073f2e04bc9ae1cf757e895f37fafecaa3bb011/platform_poc/data/store_factories.py#L51-L66):
```
S3ClientDol(
    aws_access_key_id=aws_access_key_id,
    aws_secret_access_key=aws_secret_access_key,
    endpoint_url=endpoint_url,
)
```

Reply:
Alright. In that case I think we should add a `S3Store` class to a `s3dol.s3_store` (maybe rename it to `s3dol.stores` so we can instantiate a store with the creds directly instead of having to manually instantiate the client. A similar example is in mongodol with [`MongoStore`](https://github.com/i2mint/mongodol/blob/58566027783208a1ba86d79530678735ff968314/mongodol/stores.py#L115-L121). Note that the client, db and collection are instantiated using `host`, `db_name` and `collection_name` but it is optimized with `lru_cache` so they are only created once for the same config.

Reply:
Added S3Store that must be instantiated with creds directly. Exception will be thrown if not provided.
```
    from s3dol.store import S3Store

    s3_store = S3Store(
        bucket_name='test-bucket',
        aws_access_key_id=aws_access_key_id,
        aws_secret_access_key=aws_secret_access_key,
        endpoint_url=endpoint_url,
    )
```